## ChatNS

In [42]:
from collections import deque, defaultdict
from tqdm import tqdm


def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid


def find_positions(grid):
    start = None
    end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end


def bfs(grid, start, end, cheated_pos=None):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    visited = set([start])
    queue = deque([(start, 0)])  # (position, time)

    def is_walkable(x, y):
        within_bounds = 0 <= x < len(grid) and 0 <= y < len(grid[0])
        if not within_bounds:
            return False
        if (x, y) == end or grid[x][y] == ".":
            return True
        if grid[x][y] == "#" and cheated_pos and (x, y) in cheated_pos:
            return True
        return False

    while queue:
        (x, y), t = queue.popleft()

        if (x, y) == end:
            return t

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if is_walkable(nx, ny) and ((nx, ny) not in visited):
                visited.add((nx, ny))
                queue.append(((nx, ny), t + 1))

    return float("inf")


def find_cheats(grid):
    start, end = find_positions(grid)
    initial_time = bfs(grid, start, end)

    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    time_savings = defaultdict(int)
    used_walls = set()  # Track walls used in any cheat

    for x in tqdm(range(len(grid))):
        for y in range(len(grid[0])):
            if grid[x][y] == "#":
                continue

            for dx, dy in directions:
                for cheat_steps in range(1, 3):  # Cheat for 1 or 2 steps

                    cheated_positions = []
                    wall_positions = set()
                    for step in range(1, cheat_steps + 1):
                        cheat_x = x + dx * step
                        cheat_y = y + dy * step
                        if 0 <= cheat_x < len(grid) and 0 <= cheat_y < len(grid[0]):
                            cheated_positions.append((cheat_x, cheat_y))
                            if grid[cheat_x][cheat_y] == "#":
                                wall_positions.add((cheat_x, cheat_y))

                    if all(grid[cx][cy] == "#" for cx, cy in cheated_positions):
                        beyond_x = x + dx * (cheat_steps + 1)
                        beyond_y = y + dy * (cheat_steps + 1)
                        not_used_before = wall_positions.isdisjoint(used_walls)
                        if (
                            not_used_before
                            and 0 <= beyond_x < len(grid)
                            and 0 <= beyond_y < len(grid[0])
                            and (
                                grid[beyond_x][beyond_y] == "."
                                or (beyond_x, beyond_y) == end
                            )
                        ):

                            new_time = bfs(grid, start, end, set(cheated_positions))
                            saved_time = initial_time - new_time
                            if saved_time > 0:
                                time_savings[saved_time] += 1
                                used_walls.update(wall_positions)  # Add to used walls

    return time_savings


racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
cheats = find_cheats(grid)
for saving, count in sorted(cheats.items()):
    print(
        f"There {'is' if count == 1 else 'are'} {count} cheat{'s' if count > 1 else ''} that save {saving} picosecond{'s' if saving > 1 else ''}."
    )

100%|██████████| 15/15 [00:00<00:00, 2304.90it/s]

There are 14 cheats that save 2 picoseconds.
There are 14 cheats that save 4 picoseconds.
There are 2 cheats that save 6 picoseconds.
There are 4 cheats that save 8 picoseconds.
There are 2 cheats that save 10 picoseconds.
There are 3 cheats that save 12 picoseconds.
There is 1 cheat that save 20 picoseconds.
There is 1 cheat that save 36 picoseconds.
There is 1 cheat that save 38 picoseconds.
There is 1 cheat that save 40 picoseconds.
There is 1 cheat that save 64 picoseconds.


In [43]:
with open('./Data/Day 20/input.txt') as file:
    racetrack = file.read()


grid = parse_map(racetrack)
cheats = find_cheats(grid)
for saving, count in sorted(cheats.items()):
    print(
        f"There {'is' if count == 1 else 'are'} {count} cheat{'s' if count > 1 else ''} that save {saving} picosecond{'s' if saving > 1 else ''}."
    )

100%|██████████| 141/141 [01:14<00:00,  1.89it/s]

There are 979 cheats that save 2 picoseconds.
There are 979 cheats that save 4 picoseconds.
There are 293 cheats that save 6 picoseconds.
There are 459 cheats that save 8 picoseconds.
There are 198 cheats that save 10 picoseconds.
There are 321 cheats that save 12 picoseconds.
There are 143 cheats that save 14 picoseconds.
There are 230 cheats that save 16 picoseconds.
There are 102 cheats that save 18 picoseconds.
There are 182 cheats that save 20 picoseconds.
There are 80 cheats that save 22 picoseconds.
There are 143 cheats that save 24 picoseconds.
There are 58 cheats that save 26 picoseconds.
There are 108 cheats that save 28 picoseconds.
There are 52 cheats that save 30 picoseconds.
There are 111 cheats that save 32 picoseconds.
There are 56 cheats that save 34 picoseconds.
There are 103 cheats that save 36 picoseconds.
There are 42 cheats that save 38 picoseconds.
There are 77 cheats that save 40 picoseconds.
There are 33 cheats that save 42 picoseconds.
There are 73 cheats that

In [47]:
# Calculate and print total number of cheat instances saving more than 100 picoseconds
total_savings_over_100 = sum(count for saving, count in cheats.items() if saving >= 100)
print(
    f"\nTotal number of cheats that save more than 100 picoseconds: {total_savings_over_100}"
)


Total number of cheats that save more than 100 picoseconds: 1367


In [159]:
from collections import deque, defaultdict


def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid


def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([start])
    distances = {start: 0}

    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (
                0 <= nx < len(grid)
                and 0 <= ny < len(grid[0])
                and grid[nx][ny] in valid_chars
                and (nx, ny) not in distances
            ):
                distances[(nx, ny)] = distances[(x, y)] + 1
                queue.append((nx, ny))

    return distances


def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end


def generate_rhombus_cheats(grid, start_to_coord):
    max_cheat_steps = 20  # Maximum cheats allowed steps through walls
    possible_cheats = defaultdict(dict)

    for sx, sy in start_to_coord:
        # Consider rhombus points within a maximum Manhattan distance of 20
        for distance in range(1, max_cheat_steps + 1):
            for dx in range(-distance, distance + 1):
                dy = distance - abs(dx)
                # Explore both positive and negative dy
                for dy_try in [dy, -dy]:
                    cx, cy = sx + dx, sy + dy_try
                    if 0 <= cx < len(grid) and 0 <= cy < len(grid[0]):
                        if grid[cx][cy] in (".", "E"):  # Valid end positions
                            cheat_length = distance
                            possible_cheats[(sx, sy)][(cx, cy)] = cheat_length

    return possible_cheats


def evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats):
    original_time = start_to_coord[end]
    cheat_savings = {}

    for start_coord, exits in possible_cheats.items():
        for exit_coord, cheat_length in exits.items():
            start_time = start_to_coord[start_coord]
            end_time = coord_to_end.get(exit_coord, float("inf"))
            if end_time != float("inf"):
                shortcut_time = start_time + cheat_length + end_time
                if shortcut_time < original_time:
                    saving = original_time - shortcut_time
                    cheat_savings[(start_coord, exit_coord)] = saving

    return cheat_savings


def summarize_and_print_cheat_savings(cheat_savings):
    # Use a defaultdict to count cheats by the amount of savings
    savings_count = defaultdict(int)

    for start_end_pair, saving in cheat_savings.items():
        if saving >= 50:
            savings_count[saving] += 1

    # Print the summary in the desired format
    for saving, count in sorted(savings_count.items()):
        print(f"There are {count} cheats that save {saving} picoseconds.")


racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {".", "E"})
coord_to_end = bfs_from_point(grid, end, {".", "S"})
possible_cheats = generate_rhombus_cheats(grid, start_to_coord)
cheat_savings = evaluate_cheats(
    grid, start, end, start_to_coord, coord_to_end, possible_cheats
)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(
    cheat_savings.items(), key=lambda x: x[1], reverse=True
):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")

summarize_and_print_cheat_savings(cheat_savings)

Evaluated Cheat Paths and Savings:
Cheat from (3, 1) to (7, 3) saves 76 picoseconds.
Cheat from (3, 1) to (7, 4) saves 76 picoseconds.
Cheat from (3, 1) to (7, 5) saves 76 picoseconds.
Cheat from (3, 1) to (8, 3) saves 74 picoseconds.
Cheat from (2, 1) to (7, 3) saves 74 picoseconds.
Cheat from (2, 1) to (7, 4) saves 74 picoseconds.
Cheat from (2, 1) to (7, 5) saves 74 picoseconds.
Cheat from (3, 1) to (9, 1) saves 72 picoseconds.
Cheat from (3, 1) to (9, 2) saves 72 picoseconds.
Cheat from (3, 1) to (9, 3) saves 72 picoseconds.
Cheat from (2, 1) to (8, 3) saves 72 picoseconds.
Cheat from (1, 1) to (7, 3) saves 72 picoseconds.
Cheat from (1, 1) to (7, 4) saves 72 picoseconds.
Cheat from (1, 1) to (7, 5) saves 72 picoseconds.
Cheat from (1, 2) to (7, 3) saves 72 picoseconds.
Cheat from (1, 2) to (7, 4) saves 72 picoseconds.
Cheat from (1, 2) to (7, 5) saves 72 picoseconds.
Cheat from (1, 3) to (7, 3) saves 72 picoseconds.
Cheat from (1, 3) to (7, 4) saves 72 picoseconds.
Cheat from (1, 

In [160]:
with open("./Data/Day 20/input.txt") as file:
    racetrack = file.read()


grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {".", "E"})
coord_to_end = bfs_from_point(grid, end, {".", "S"})
possible_cheats = generate_rhombus_cheats(grid, start_to_coord)
cheat_savings = evaluate_cheats(
    grid, start, end, start_to_coord, coord_to_end, possible_cheats
)
count = 0
for start_end_pair, saving in cheat_savings.items():
    if saving >= 100:
        count += 1
print(count)

1006850


# Partly ChatNS history due to closure of browser

## System

You are an excellent python programmer with 10 years of experience. In this time you have excelled at solving problems with different algoritms, and are great at solving advent of code.

## User

The Historians are quite pixelated again. This time, a massive, black building looms over you - you're right outside the CPU!

While The Historians get to work, a nearby program sees that you're idle and challenges you to a race. Apparently, you've arrived just in time for the frequently-held race condition festival!

The race takes place on a particularly long and twisting code path; programs compete to see who can finish in the fewest picoseconds. The winner even gets their very own mutex!

They hand you a map of the racetrack (your puzzle input). For example:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
The map consists of track (.) - including the start (S) and end (E) positions (both of which also count as track) - and walls (#).

When a program runs through the racetrack, it starts at the start position. Then, it is allowed to move up, down, left, or right; each such move takes 1 picosecond. The goal is to reach the end position as quickly as possible. In this example racetrack, the fastest time is 84 picoseconds.

Because there is only a single path from the start to the end and the programs all go the same speed, the races used to be pretty boring. To make things more interesting, they introduced a new rule to the races: programs are allowed to cheat.

The rules for cheating are very strict. Exactly once during a race, a program may disable collision for up to 2 picoseconds. This allows the program to pass through walls as if they were regular track. At the end of the cheat, the program must be back on normal track again; otherwise, it will receive a segmentation fault and get disqualified.

So, a program could complete the course in 72 picoseconds (saving 12 picoseconds) by cheating for the two moves marked 1 and 2:

###############
#...#...12....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
Or, a program could complete the course in 64 picoseconds (saving 20 picoseconds) by cheating for the two moves marked 1 and 2:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...12..#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
This cheat saves 38 picoseconds:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.####1##.###
#...###.2.#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
This cheat saves 64 picoseconds and takes the program directly to the end:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..21...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
Each cheat has a distinct start position (the position where the cheat is activated, just before the first move that is allowed to go through walls) and end position; cheats are uniquely identified by their start position and end position.

In this example, the total number of cheats (grouped by the amount of time they save) are as follows:

There are 14 cheats that save 2 picoseconds.
There are 14 cheats that save 4 picoseconds.
There are 2 cheats that save 6 picoseconds.
There are 4 cheats that save 8 picoseconds.
There are 2 cheats that save 10 picoseconds.
There are 3 cheats that save 12 picoseconds.
There is one cheat that saves 20 picoseconds.
There is one cheat that saves 36 picoseconds.
There is one cheat that saves 38 picoseconds.
There is one cheat that saves 40 picoseconds.
There is one cheat that saves 64 picoseconds.
You aren't sure what the conditions of the racetrack will be like, so to give yourself as many options as possible, you'll need a list of the best cheats. How many cheats would save you at least 100 picoseconds?

Your puzzle answer was 1367.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The programs seem perplexed by your list of cheats. Apparently, the two-picosecond cheating rule was deprecated several milliseconds ago! The latest version of the cheating rule permits a single cheat that instead lasts at most 20 picoseconds.

Now, in addition to all the cheats that were possible in just two picoseconds, many more cheats are possible. This six-picosecond cheat saves 76 picoseconds:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#1#####.#.#.###
#2#####.#.#...#
#3#####.#.###.#
#456.E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
Because this cheat has the same start and end positions as the one above, it's the same cheat, even though the path taken during the cheat is different:

###############
#...#...#.....#
#.#.#.#.#.###.#
#S12..#.#.#...#
###3###.#.#.###
###4###.#.#...#
###5###.#.###.#
###6.E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
Cheats don't need to use all 20 picoseconds; cheats can last any amount of time up to and including 20 picoseconds (but can still only end when the program is on normal track). Any cheat time not used is lost; it can't be saved for another cheat later.

You'll still need a list of the best cheats, but now there are even more to choose between. Here are the quantities of cheats in this example that save 50 picoseconds or more:

There are 32 cheats that save 50 picoseconds.
There are 31 cheats that save 52 picoseconds.
There are 29 cheats that save 54 picoseconds.
There are 39 cheats that save 56 picoseconds.
There are 25 cheats that save 58 picoseconds.
There are 23 cheats that save 60 picoseconds.
There are 20 cheats that save 62 picoseconds.
There are 19 cheats that save 64 picoseconds.
There are 12 cheats that save 66 picoseconds.
There are 14 cheats that save 68 picoseconds.
There are 12 cheats that save 70 picoseconds.
There are 22 cheats that save 72 picoseconds.
There are 4 cheats that save 74 picoseconds.
There are 3 cheats that save 76 picoseconds.
Find the best cheats using the updated cheating rules. How many cheats would save you at least 100 picoseconds?

## Assistant

To tackle this problem, we'll utilize the following approach:

1. **Parse the Input Map**: Convert the input ASCII map into a usable format like a grid, where each cell represents a wall (`#`), open track (`.`), the start (`S`), or the end (`E`). We'll start by finding the coordinates of the start and end points.

2. **Breadth-First Search (BFS) for Shortest Path**: Perform a BFS to find the shortest path from start to end without any cheating. This will be our baseline path and will indicate the total number of cells traveled.

3. **Identify Cheating Opportunities**: For each possible position and direction in the grid, simulate the cheat by assuming a walk through walls for up to 20 moves, and count both how long the cheat takes and how much time it saves. 

4. **Count Cheats with Savings >= 100 Picoseconds**: For each cheating opportunity, compute the total time saved compared to the no-cheat path and count the ones that save at least 100 picoseconds.

Here's a rough implementation outline in Python:

```python
from collections import deque

# Define directions: (row_change, col_change)
DIRECTIONS = [(0, 1), (1, 0), (0, -1), (-1, 0)]

def parse_map(racetrack):
    # Parse the map into a grid and find start (S) and end (E) positions
    grid = []
    start, end = None, None
    for r, line in enumerate(racetrack):
        row = []
        for c, char in enumerate(line):
            if char == 'S':
                start = (r, c)
                row.append('.')
            elif char == 'E':
                end = (r, c)
                row.append('.')
            else:
                row.append(char)
        grid.append(row)
    return grid, start, end

def bfs_no_cheat(grid, start, end):
    # Standard BFS to find shortest path from start to end without cheating
    queue = deque([(start[0], start[1], 0)])  # (row, col, time)
    visited = set()
    visited.add(start)
    
    while queue:
        r, c, time = queue.popleft()
        
        if (r, c) == end:
            return time
        
        for dr, dc in DIRECTIONS:
            nr, nc = r + dr, c + dc
            
            if (0 <= nr < len(grid) and 0 <= nc < len(grid[0]) and
                grid[nr][nc] == '.' and (nr, nc) not in visited):
                visited.add((nr, nc))
                queue.append((nr, nc, time + 1))
    
    return float('inf')  # If no path is found

def bfs_with_cheat(grid, start, end, max_cheat_moves=20):
    # Find all cheats that save >= 100 picoseconds
    original_no_cheat_time = bfs_no_cheat(grid, start, end)
    count_cheats = 0

    for sr in range(len(grid)):
        for sc in range(len(grid[0])):
            if grid[sr][sc] != '.':
                continue
            
            # Try starting a cheat from (sr, sc)
            for dr, dc in DIRECTIONS:
                r, c = sr, sc
                time_spent_cheating = 0
                cheated_path = []
    
                while time_spent_cheating < max_cheat_moves:
                    r, c = r + dr, c + dc
                    if not (0 <= r < len(grid) and 0 <= c < len(grid[0])):
                        break # out of bounds
    
                    if grid[r][c] == '.':
                        break # back to normal path
            
                    cheated_path.append((r, c))
                    time_spent_cheating += 1

                    # Attempt to reach the end normally
                    if grid[r][c] == '.':
                        # Make a temporary copy of grid with the cheat path as open
                        temp_grid = [row[:] for row in grid]
                        for cr, cc in cheated_path:
                            temp_grid[cr][cc] = '.'
                        
                        total_time = bfs_no_cheat(temp_grid, start, end)
                        saved_time = original_no_cheat_time - total_time
                        
                        if saved_time >= 100:
                            count_cheats += 1
                        break
    return count_cheats

def solve(racetrack):
    grid, start, end = parse_map(racetrack.splitlines())
    return bfs_with_cheat(grid, start, end)

# Example racetrack
racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

print(solve(racetrack))
```
Ensure you adjust the traversal logic and implementation based on the shape and specifics of your input map and accurately analyzing all possible cheats. Since the problem asks for performing a significant number of cheat scenarios, optimizations and smart pruning might be required on the test scale.

## User

```python
from collections import deque, defaultdict


def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid


def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end


def bfs(grid, start, targets, prohibited=None):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([(start, 0)])
    visited = set([start])

    while queue:
        (x, y), t = queue.popleft()

        if (x, y) in targets:
            return t

        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (
                0 <= nx < len(grid)
                and 0 <= ny < len(grid[0])
                and (grid[nx][ny] in (".", "S", "E") or (nx, ny) in targets)
                and ((nx, ny) not in visited)
                and (prohibited is None or (nx, ny) not in prohibited)
            ):

                visited.add((nx, ny))
                queue.append(((nx, ny), t + 1))

    return float("inf")


def generate_rhombus(grid, x, y):
    max_distance = 20  # Maximum allowed cheat distance
    shortcuts = []

    # Nested loop to create a diamond shape
    for distance in range(1, max_distance + 1):
        for dx in range(-distance, distance + 1):
            dy = distance - abs(dx)
            for dy_try in [dy, -dy]:
                cx, cy = x + dx, y + dy_try
                if 0 <= cx < len(grid) and 0 <= cy < len(grid[0]):
                    if grid[cx][cy] == ".":
                        shortcuts.append((cx, cy))
                    elif grid[cx][cy] == "E":
                        shortcuts.append((cx, cy))
                    if grid[cx][cy] == "#":
                        break

    return shortcuts


def calculate_savings(grid):
    start, end = find_positions(grid)
    original_time = bfs(grid, start, {end})

    shortcut_savings = defaultdict(int)

    # Generate shortcuts and evaluate their time savings
    for x in range(len(grid)):
        for y in range(len(grid[0])):
            if grid[x][y] not in (".", "S", "E"):
                continue

            for ex, ey in generate_rhombus(grid, x, y):
                if (ex, ey) == (x, y):
                    continue

                sc_start = bfs(grid, start, {(x, y)})  # Path to shortcut start
                sc_end = bfs(grid, (ex, ey), {end})  # Path from shortcut end

                # Calculate the length of the shortcut (Manhattan distance)
                shortcut_length = abs(ex - x) + abs(ey - y)

                # Calculate the saved time
                if sc_start != float("inf") and sc_end != float("inf"):
                    total_time_with_shortcut = sc_start + sc_end + shortcut_length
                    saved_time = original_time - total_time_with_shortcut
                    if saved_time > 0:
                        key = ((x, y), (ex, ey))
                        shortcut_savings[key] = max(shortcut_savings[key], saved_time)

    return shortcut_savings


def count_shortcuts_with_savings(shortcuts, min_savings):
    counts = defaultdict(int)
    for saving in shortcuts.values():
        if saving >= min_savings:
            counts[saving] += 1
    return counts


racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
shortcut_savings = calculate_savings(grid)

# # Display results focusing on keys and their savings time
print("Shortcuts with significant savings:")
for ((sx, sy), (ex, ey)), saving in sorted(
    shortcut_savings.items(), key=lambda x: x[1], reverse=True
):
    if saving >= 50:
        print(f"Shortcut from ({sx}, {sy}) to ({ex}, {ey}) saves: {saving} picoseconds")

# Calculate and output the number of shortcuts saving 50 picoseconds or more
min_savings = 50
shortcut_count = count_shortcuts_with_savings(shortcut_savings, min_savings)
for saving, count in sorted(shortcut_count.items()):
    print(f"Number of shortcuts saving {count} or more picoseconds: {saving}")
```
Currently you have created this code

## Assistant

Great! It looks like you've implemented a code solution to find and evaluate shortcuts on the given racetrack, where programs can cheat by going through walls for up to 20 picoseconds. You have a function to calculate the potential savings of each possible cheat and then determine how many cheats save a minimum amount of time.

Let's walk through what the code is doing and make sure it aligns with your objectives:

- **`parse_map()`**: Converts the racetrack input into a 2D grid, which is easier to work with programmatically.
- **`find_positions()`**: Finds and returns the start 'S' and end 'E' positions in the grid.
- **`bfs()`**: A BFS function that calculates the shortest path from a start position to any of the `targets`, avoiding specified prohibited cells.
- **`generate_rhombus()`**: Creates a rhombus-shaped pattern (bounded by 20 cells) from any given position. This pattern represents the possible cells that can be reached with a cheat.
- **`calculate_savings()`**: Computes the time saved by each possible shortcut. It finds partial paths from start to cheat start, cheat end to end, and then calculates savings by comparing to the original path time.
- **`count_shortcuts_with_savings()`**: Counts how many cheats result in a time saving over a specified threshold.

The main script then assembles these components, calculates the shortcut savings, and outputs the number of shortcuts saving significant time.

Here's how you might refine the results and output it clearly:

```python
grid = parse_map(racetrack)
shortcut_savings = calculate_savings(grid)

# Display results focusing on keys and their savings time
print("Shortcuts with significant savings:")
for ((sx, sy), (ex, ey)), saving in sorted(
    shortcut_savings.items(), key=lambda x: x[1], reverse=True
):
    if saving >= 50:
        print(f"Shortcut from ({sx}, {sy}) to ({ex}, {ey}) saves: {saving} picoseconds")

# Calculate and output the number of shortcuts saving a significant amount of time
min_savings = 50
shortcut_count = count_shortcuts_with_savings(shortcut_savings, min_savings)
print(f"Total number of shortcuts saving at least {min_savings} picoseconds:")
for saving, count in sorted(shortcut_count.items(), reverse=True):
    print(f"Savings: {saving} picoseconds, Count: {count}")
```

### Considerations:

1. **Efficiency**: 
   - Ensure that the BFS paths are explored optimally, especially since you're allowing up to 20-cell deviations for cheats.
   - If there are larger maps, consider implementations that optimize BFS checks to avoid redundant calculations or use memoization where applicable.

2. **Validation**:
   - Verify the grid boundaries in `generate_rhombus()` because incorrect conditions might lead to incorrect shortcut paths.
   - Ensure the paths calculated actually respect the game's moving rules, particularly returning to a valid track cell after the cheat.

3. **Input Handling**:
   - If your input's dimensions might change or you have multiple input cases, it might be wise to add input validation or prepare the function for extensions as needed.

This should successfully give you the list of significant shortcut savings and a count of those that exceed your defined threshold. Let me know if there are any specific areas or edge cases you'd like to explore further!

## User

Okay new approach.
First create a dictionary with the key the coordinate and the value being the shortest score from the coordinate till the finish. Next to this create a second dictionary with the key the coordinate and the value being the shortest score till the coordinate from the start. 
After having built these dictionary start cheating. This means that for every coordinate (key) you try to find all possible next level keys that are less then 20 steps away passing only through #. Using this approach start calculating the new route speed by taking the value of the first dictionary for the end of the cheet and taking the value of the second dictionary for the time of the start of the shortcut, increase this with the length through # as the time taken cheating

## Assistant

This is a clever approach that optimizes the computation by precalculating the shortest paths from the start and to the end of the racetrack for all coordinates. This way, you can focus on evaluating possible cheating shortcuts, leveraging precomputed distances to rapidly evaluate possible cheat paths.

Here's how you can implement this approach:

### Step 1: Precompute Shortest Paths

1. **From Start to Each Coordinate**: Use BFS to compute the shortest paths from the start position (`S`) to all other coordinates in the grid. Store these distances in a dictionary (`start_to_coord`).

2. **From Each Coordinate to End**: Similarly, perform BFS from the end position (`E`) to all other coordinates but in reverse (you can think of this like a BFS where you start at `E` and explore backwards). Store these distances in another dictionary (`coord_to_end`).

### Step 2: Evaluate Cheating Shortcuts

1. For each possible start of a cheat (each open track cell `.`), determine all possible end positions reachable within 20 steps, passing only through walls `#`.

2. Use the `start_to_coord` and `coord_to_end` dictionaries to compute the total time of the shortcut route and compare it with the original time.

3. Track the time savings for the cheat paths that result in a valid solution.

Here's an implementation outline:

```python
from collections import deque, defaultdict

def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid

def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([(start, 0)])
    distances = {start: 0}
    
    while queue:
        (x, y), dist = queue.popleft()
        
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and
                    grid[nx][ny] in valid_chars and (nx, ny) not in distances):
                distances[(nx, ny)] = dist + 1
                queue.append(((nx, ny), dist + 1))
                
    return distances

def find_shortest_paths(grid):
    start, end = find_positions(grid)
    start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
    coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
    return start_to_coord, coord_to_end

def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end

def evaluate_cheats(grid, start_to_coord, coord_to_end):
    original_time = start_to_coord[end]
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    cheat_times = defaultdict(lambda: float('inf'))

    for (sx, sy) in start_to_coord:
        if grid[sx][sy] != '.':
            continue
        for dx, dy in directions:
            cx, cy = sx, sy
            time_cheat = 0
            while time_cheat < 20:
                cx += dx
                cy += dy
                if not (0 <= cx < len(grid) and 0 <= cy < len(grid[0])):
                    break  # Out of bounds
                
                if grid[cx][cy] == '.':
                    end_time = coord_to_end.get((cx, cy), float('inf'))
                    if end_time != float('inf'):
                        shortcut_time = start_to_coord[(sx, sy)] + time_cheat + 1 + end_time
                        if shortcut_time < original_time:
                            cheat_times[((sx, sy), (cx, cy))] = min(cheat_times[((sx, sy), (cx, cy))], original_time - shortcut_time)
                    break
                
                if grid[cx][cy] == '#':
                    time_cheat += 1
   
    return cheat_times

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start_to_coord, coord_to_end = find_shortest_paths(grid)
cheat_times = evaluate_cheats(grid, start_to_coord, coord_to_end)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_times.items(), key=lambda x: x[1], reverse=True):
    print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")
```

### Explanation:

- **`bfs_from_point`**: Performs a BFS search to calculate shortest paths either from the start or from the end to all key positions.

- **`find_shortest_paths`**: Utilizes `bfs_from_point` to compute paths from start and to end.

- **`evaluate_cheats`**: Iterates through potential cheat starting points, checking each direction for valid cheat paths while updating the shortest time if a cheat results in a quicker path.

This solution should be quite efficient as much of the heavy lifting is done in the BFS precomputation steps, and cheating evaluations are greatly simplified with precomputed path lengths.

## User

You are mixing the indexes of the dictionary do all coordinates as x,y

## Assistant

Got it! I apologize for any confusion. Let's make the necessary adjustments and ensure that all coordinates are consistently represented as `(x, y)` in the dictionaries and throughout the calculations. Here's the updated implementation:

### Updated Implementation

```python
from collections import deque, defaultdict

def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid

def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([start])
    distances = {start: 0}
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and
                    grid[nx][ny] in valid_chars and (nx, ny) not in distances):
                distances[(nx, ny)] = distances[(x, y)] + 1
                queue.append((nx, ny))
                
    return distances

def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end

def evaluate_cheats(grid, start, end, start_to_coord, coord_to_end):
    original_time = start_to_coord[end]
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    cheat_savings = defaultdict(lambda: float('inf'))

    for (sx, sy) in start_to_coord:
        if grid[sx][sy] != '.':
            continue
        for dx, dy in directions:
            cx, cy = sx, sy
            time_cheat = 0
            while time_cheat < 20:
                cx += dx
                cy += dy
                if not (0 <= cx < len(grid) and 0 <= cy < len(grid[0])):
                    break
                
                if grid[cx][cy] == '.':
                    end_time = coord_to_end.get((cx, cy), float('inf'))
                    if end_time != float('inf'):
                        shortcut_time = start_to_coord[(sx, sy)] + time_cheat + end_time
                        if shortcut_time < original_time:
                            cheat_savings[((sx, sy), (cx, cy))] = max(cheat_savings[((sx, sy), (cx, cy))], original_time - shortcut_time)
                    break
                
                if grid[cx][cy] == '#':
                    time_cheat += 1
   
    return cheat_savings

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
cheat_savings = evaluate_cheats(grid, start, end, start_to_coord, coord_to_end)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_savings.items(), key=lambda x: x[1], reverse=True):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")
```

### Key Adjustments:

1. **Coordinate Representation**: Ensured all coordinates consistently use `(x, y)` across the dictionaries and loop logic.

2. **Evaluating Cheating Paths**: Uses nested loops to check each coordinate if it's eligible to begin a cheat path and calculates the possible savings for a direct path of up to 20 `#` cells.

3. **Savings Calculation**: Calculates shortcuts by summing parts�distance from start to shortcut start, cheat length, and from cheat end to the actual end.

4. **Output**: Displays cheats that result in positive time savings only, which are those that complete the course faster than without a cheat.

### Note:
- Ensure the input `racetrack` string accurately represents the boundaries and paths to avoid unintended index errors.
- The `parse_map()`, `find_positions()`, and BFS helper functions are designed to handle grid operations, ensuring flexibility and debugging ease.
- Further optimizations might involve pruning potential cheats based on early bounding or heuristic assessments on long paths.

## User

can you extend the check in checking if it is a cheat to include the S and E locations? Only walls can't be start or end points

## Assistant

Certainly! To extend the functionality so that both the start (`S`) and end (`E`) can also be considered valid start and end points for a cheat (rather than only open path cells `.`), we should ensure that these positions are also checked whenever we are calculating the potential cheat paths. Here's how you can implement this:

### Updated Code

```python
from collections import deque, defaultdict

def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid

def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([start])
    distances = {start: 0}
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and
                    grid[nx][ny] in valid_chars and (nx, ny) not in distances):
                distances[(nx, ny)] = distances[(x, y)] + 1
                queue.append((nx, ny))
                
    return distances

def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end

def evaluate_cheats(grid, start, end, start_to_coord, coord_to_end):
    original_time = start_to_coord[end]
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    cheat_savings = defaultdict(lambda: float('inf'))

    for (sx, sy) in start_to_coord:
        if grid[sx][sy] in ('#'):  # Now only wall locations are invalid
            continue
        for dx, dy in directions:
            cx, cy = sx, sy
            time_cheat = 0
            while time_cheat < 20:
                cx += dx
                cy += dy
                if not (0 <= cx < len(grid) and 0 <= cy < len(grid[0])):
                    break  # Out of bounds
                
                if grid[cx][cy] in ('.', 'E'):  # We want to allow end positions
                    end_time = coord_to_end.get((cx, cy), float('inf'))
                    if end_time != float('inf'):
                        shortcut_time = start_to_coord[(sx, sy)] + time_cheat + end_time
                        if shortcut_time < original_time:
                            cheat_savings[((sx, sy), (cx, cy))] = max(cheat_savings[((sx, sy), (cx, cy))], original_time - shortcut_time)
                    break
                
                if grid[cx][cy] == '#':
                    time_cheat += 1
   
    return cheat_savings

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
cheat_savings = evaluate_cheats(grid, start, end, start_to_coord, coord_to_end)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_savings.items(), key=lambda x: x[1], reverse=True):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")
```

### Key Changes:
- **Valid Start and End Cells**: Now, only `#` cells are considered invalid for starting or ending a cheat path, allowing `S` and `E` to be valid start/end positions.

- **Cheat Evaluation**: When evaluating a cheat path, all directions are considered from any cell except `#`.

This setup enables comprehensive checking for valid shortcuts, facilitating strategic path cheating to minimize overall time effectively.

## User

Okay change the cheat finding algorithm. I want to add a step before evaluating namely finding all cheats possible from any coordinate in the map of start_to_coord do a rhumbus creation around each point with the length of the center to a side being 20. Search for all points within the rumbus that have a . and create this as possible ending coordinated, which you add add it as a possible exit_node including the length in a new dictionary which has the key the coordinate of the shortcut then as a value a dictionary with the keys the exit_nodes and the values the shortcut length

## Assistant

To implement the updated cheat-finding algorithm, we'll follow these steps:

1. **Precompute Shortest Paths**: First, calculate the shortest paths from the start and to the end for all coordinates as before.

2. **Generate Rhombus for Each Coordinate**: For each coordinate in `start_to_coord`, generate a rhombus with a Manhattan distance of up to 20. This will allow us to check possible end points for cheats originating from each coordinate.

3. **Identify Valid Cheat Endpoints**: Within the rhombus, identify all valid cheat end points (i.e., coordinates with `.` or `E`) and calculate the cheat length (distance through walls).

4. **Store Cheat Information**: Create a new dictionary that stores, for each start coordinate, a dictionary of possible end coordinates and the corresponding cheat lengths.

Here's how you can implement this in Python:

```python
from collections import deque, defaultdict

def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid

def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([start])
    distances = {start: 0}
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and
                    grid[nx][ny] in valid_chars and (nx, ny) not in distances):
                distances[(nx, ny)] = distances[(x, y)] + 1
                queue.append((nx, ny))
                
    return distances

def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end

def generate_cheats(grid, start_to_coord):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    possible_cheats = defaultdict(dict)

    for (sx, sy) in start_to_coord:
        if grid[sx][sy] == '#':  # Only non-wall cells can be cheat starting points
            continue

        for dx, dy in directions:
            steps = 0
            cx, cy = sx, sy
            cheat_length_to_end = 0

            # Generate a rhombus path
            while steps < 20:
                cx += dx
                cy += dy
                if not (0 <= cx < len(grid) and 0 <= cy < len(grid[0])):
                    break  # Out of bounds
                
                if grid[cx][cy] == '#':
                    cheat_length_to_end += 1
                elif grid[cx][cy] in ('.', 'E'):
                    possible_cheats[(sx, sy)][(cx, cy)] = cheat_length_to_end
                    break  # Stop if valid end point is reached
                else:
                    break  # Stop if any other invalid cell is reached

                steps += 1

    return possible_cheats

def evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats):
    original_time = start_to_coord[end]
    cheat_savings = defaultdict(lambda: float('inf'))

    for (start_coord, exits) in possible_cheats.items():
        for (exit_coord, cheat_length) in exits.items():
            start_time = start_to_coord[start_coord]
            end_time = coord_to_end.get(exit_coord, float('inf'))
            if end_time != float('inf'):
                shortcut_time = start_time + cheat_length + end_time
                if shortcut_time < original_time:
                    saving = original_time - shortcut_time
                    cheat_savings[(start_coord, exit_coord)] = saving

    return cheat_savings

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
possible_cheats = generate_cheats(grid, start_to_coord)
cheat_savings = evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_savings.items(), key=lambda x: x[1], reverse=True):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")
```

### Explanation:

- **Cheat Generation**: For each start position in `start_to_coord`, generate a rhombus of potential cheat endpoints. The endpoints can be anything but walls and out-of-bounds cells.

- **Cheat Evaluation**: Use the precomputed dictionaries (`start_to_coord` and `coord_to_end`) to evaluate possible cheats and calculate their savings in time. Only cheats with positive savings are stored and printed.

This approach ensures a comprehensive check for possible cheat paths, supporting efficient evaluation by leveraging the precomputed paths.

## User

You ar not building a rhombus you make a cross

## Assistant

Thank you for pointing that out. To properly generate a rhombus for potential cheats, we need to consider all positions within a Manhattan distance of 20 steps from a given starting coordinate. This means exploring positions that form a diamond shape rather than a cross. Let's correct that implementation.

### Updated Implementation: Generating a Rhombus

For generating a rhombus shape, we should explore positions iteratively around a starting point, checking all coordinates in which the sum of their relative distances (in terms of both x and y changes) from the center doesn't exceed the maximum allowed steps (20 steps, in this case).

Here's how you can implement this:

```python
from collections import deque, defaultdict

def parse_map(racetrack):
    lines = racetrack.strip().split("\n")
    grid = [list(line) for line in lines]
    return grid

def bfs_from_point(grid, start, valid_chars):
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    queue = deque([start])
    distances = {start: 0}
    
    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and
                    grid[nx][ny] in valid_chars and (nx, ny) not in distances):
                distances[(nx, ny)] = distances[(x, y)] + 1
                queue.append((nx, ny))
                
    return distances

def find_positions(grid):
    start = end = None
    for r, row in enumerate(grid):
        for c, val in enumerate(row):
            if val == "S":
                start = (r, c)
            elif val == "E":
                end = (r, c)
    return start, end

def generate_rhombus_cheats(grid, start_to_coord):
    max_cheat_steps = 20  # Maximum cheats allowed steps through walls
    possible_cheats = defaultdict(dict)

    for (sx, sy) in start_to_coord:
        # Consider rhombus points within a maximum Manhattan distance of 20
        for distance in range(1, max_cheat_steps + 1):
            for dx in range(-distance, distance + 1):
                dy = distance - abs(dx)
                # Explore both positive and negative dy
                for dy_try in [dy, -dy]:
                    cx, cy = sx + dx, sy + dy_try
                    if (0 <= cx < len(grid) and 0 <= cy < len(grid[0])):
                        if grid[cx][cy] in ('.', 'E'):  # Valid end positions
                            cheat_length = distance
                            possible_cheats[(sx, sy)][(cx, cy)] = cheat_length

    return possible_cheats

def evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats):
    original_time = start_to_coord[end]
    cheat_savings = {}

    for start_coord, exits in possible_cheats.items():
        for exit_coord, cheat_length in exits.items():
            start_time = start_to_coord[start_coord]
            end_time = coord_to_end.get(exit_coord, float('inf'))
            if end_time != float('inf'):
                shortcut_time = start_time + cheat_length + end_time
                if shortcut_time < original_time:
                    saving = original_time - shortcut_time
                    cheat_savings[(start_coord, exit_coord)] = saving

    return cheat_savings

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
possible_cheats = generate_rhombus_cheats(grid, start_to_coord)
cheat_savings = evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_savings.items(), key=lambda x: x[1], reverse=True):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")
```

### Explanation:

- **Rhombus Generation**: We iterate over all potential distances (from 1 to the maximum allowed distance) and explore all coordinates within those distances in a diamond pattern around each starting point. This covers the set of coordinates with a Manhattan distance up to the maximum.

- **Cheating Path Evaluation**: Uses the generated possible cheats and precomputed path lengths to evaluate and determine the savings.

This implementation ensures a comprehensive selection and evaluates all possible cheats while adhering to the new criteria of not exceeding a given cheat length within the rhombus boundaries.

## User

Can you add a final function which transforms the cheats_saving dictionary in a bunch of prints of the following format: There are 32 cheats that save 50 picoseconds. 

## Assistant

Certainly! We can write a function that summarizes the `cheat_savings` dictionary to count how many cheats provide a given amount of time savings and print those counts in the specified format.

Here's how you can do it:

```python
def summarize_and_print_cheat_savings(cheat_savings):
    # Use a defaultdict to count cheats by the amount of savings
    savings_count = defaultdict(int)
    
    for start_end_pair, saving in cheat_savings.items():
        if saving > 0:
            savings_count[saving] += 1
    
    # Print the summary in the desired format
    for saving, count in sorted(savings_count.items()):
        print(f"There are {count} cheats that save {saving} picoseconds.")

# Given implementation parts remain unchanged in this section

racetrack = """
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

grid = parse_map(racetrack)
start, end = find_positions(grid)
start_to_coord = bfs_from_point(grid, start, {'.', 'E'})
coord_to_end = bfs_from_point(grid, end, {'.', 'S'})
possible_cheats = generate_rhombus_cheats(grid, start_to_coord)
cheat_savings = evaluate_cheats(grid, start, end, start_to_coord, coord_to_end, possible_cheats)

print("Evaluated Cheat Paths and Savings:")
for ((sx, sy), (ex, ey)), saving in sorted(cheat_savings.items(), key=lambda x: x[1], reverse=True):
    if saving > 0:
        print(f"Cheat from ({sx}, {sy}) to ({ex}, {ey}) saves {saving} picoseconds.")

# Print the summarized cheats
summarize_and_print_cheat_savings(cheat_savings)
```

### Explanation:

- **`summarize_and_print_cheat_savings`**: This function iterates over the `cheat_savings` dictionary to count and group all cheats by their picosecond savings. It then prints the summary in the desired format.
  
- **Output**: The print statements reflect how many unique cheats provide specific time savings, formatted as specified.

By calling the `summarize_and_print_cheat_savings` function after evaluating cheat paths, you can easily see how efficiently different cheats improve race times in your solution.

## User

Great job you solved the challenge

